cpp/src/cylon/serialize/table_serialize.hpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CYLON_CPP_SRC_CYLON_NET_TABLE_SERIALIZE_HPP_
+#define CYLON_CPP_SRC_CYLON_NET_TABLE_SERIALIZE_HPP_
+
+#include "cylon/table.hpp"
+#include "cylon/net/serialize.hpp"
+
+namespace cylon{
+
+class CylonTableSerializer : public TableSerializer {
+ public:
+  CylonTableSerializer(std::shared_ptr<arrow::Table> table,
+                       std::vector<int32_t> buffer_sizes,
+                       std::vector<const uint8_t *> data_buffers,
+                       arrow::BufferVector extra_buffers);
+
+  static Status Make(const std::shared_ptr<Table> &table,
+                     std::shared_ptr<TableSerializer> *serializer);
+
+  const std::vector<int32_t> &getBufferSizes() override {
+    return buffer_sizes_;
+  }
+  int getNumberOfBuffers() override {
+    return num_buffers_;
+  }
+  std::vector<int32_t> getEmptyTableBufferSizes() override;
+
+  const std::vector<const uint8_t *> &getDataBuffers() override {
+    return data_buffers_;
+  }
+
+  std::vector<int32_t> getDataTypes() override;
+
+  const arrow::BufferVector &extra_buffers() const;
+
+ private:
+
+  const std::shared_ptr<arrow::Table> table_;
+  const int32_t num_buffers_;
+  const std::vector<int32_t> buffer_sizes_;
+  const std::vector<const uint8_t *> data_buffers_;
+
+  // when there are boolean buffers with non-byte-boundary-offsets, we can not get a byte* to the
+  // data start. So, we'll have to make copy to new buffer. These new buffers will be kept here.
```

Comment:
why do we keep new buffer in a different place. can't we just keep the new buffers in data_buffers_?

Reply:
`data_buffers_` are just pointers right? 
Actually these are the copied bool buffers. There wouldn't be any tables that holds these `extra_buffers_`. They would only be created during the serializing step, and the serializer owns those buffers (untill they are consumed).

Reply:
ok, so the class member may be needed for deallocating the memory. 
But, we may delete the method extra_buffers() if we already point to this buffer from data_buffers_.

Reply:
well, I havent added a method to clear them (yet). These would be deleted when the serializer gets destroyed.
If we are shuffling we dont have to worry about it because, its already copied to a remote memory location by the time the serializer falls off.